reader_test.go

```diff
@@ -0,0 +1,239 @@
+package iscsi_test
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	iscsi "github.com/willgorman/libiscsi-go"
+	"gotest.tools/assert"
+)
+
+func TestRead(t *testing.T) {
+	seed := time.Now().UnixNano()
+	t.Logf("using seed %d", seed)
+	rnd := rand.New(rand.NewSource(seed))
+	fileName := writeTargetTempfile(t, rnd, 4*KiB)
+	file, err := os.Open(fileName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer file.Close()
+	hash := sha256.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		log.Fatal(err)
+	}
+	fileChecksum := fmt.Sprintf("%x", hash.Sum(nil))
+	t.Log("FILE CHECKSUM", fileChecksum)
+
+	device := iscsi.New(iscsi.ConnectionDetails{
+		InitiatorIQN: "iqn.2024-10.libiscsi:go",
+		TargetURL:    runTestTarget(t, fileName),
+	})
+
+	err = device.Connect()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_ = device.Disconnect()
+	}()
+
+	sreader, err := iscsi.Reader(device)
+	if err != nil {
+		t.Fatal(err)
+	}
+	log.Printf("%#v", sreader)
+	hash = sha256.New()
+	if _, err := io.Copy(hash, sreader); err != nil {
+		log.Fatal(err)
+	}
+	iscsiChecksum := fmt.Sprintf("%x", hash.Sum(nil))
+	t.Log("ISCSI CHECKSUM ", iscsiChecksum)
+	assert.Equal(t, fileChecksum, iscsiChecksum)
+}
+
+// in order to test non block aligned reads
+// we can have a file io.Reader and iscsi io.Reader and
+// read randomly sized []byte from them and assert that we always get
+// the same values from each
+func TestReadRandom(t *testing.T) {
+	seed := time.Now().UnixNano()
+	t.Logf("using seed %d", seed)
+	rnd := rand.New(rand.NewSource(seed))
+	fileName := writeTargetTempfile(t, rnd, 10*MiB)
+	file, err := os.Open(fileName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer file.Close()
+
+	device := iscsi.New(iscsi.ConnectionDetails{
+		InitiatorIQN: "iqn.2024-10.libiscsi:go",
+		TargetURL:    runTestTarget(t, fileName),
+	})
+
+	err = device.Connect()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_ = device.Disconnect()
+	}()
+
+	sreader, err := iscsi.Reader(device)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var fileErr, scsiErr error
+	var fileN, scsiN int
+	for fileErr != io.EOF && scsiErr != io.EOF {
+		n := rnd.Intn(32 * KiB)
+		fileBytes := make([]byte, n)
+		scsiBytes := make([]byte, n)
+		fileN, fileErr = file.Read(fileBytes)
+		if fileErr != nil && fileErr != io.EOF {
+			t.Fatal(fileErr)
+		}
+		retry.Do(func() error {
+			scsiN, scsiErr = sreader.Read(scsiBytes)
+			return scsiErr
+		}, retry.RetryIf(func(err error) bool {
+			if err != nil && strings.Contains(err.Error(), "Poll failed") {
+				return true
+			}
+			return false
```

Comment:
Yes, it happens fairly often and it _seems_ like retrying is always ultimately successful but I've been hesitant to add retrying within the api since I don't understand what the cause is.  Poll is a syscall watching the file descriptor for the tcp socket for events but I don't know what a failure means (and it doesn't seem to just be a timeout)  When it succeeds on a retry I believe it is doing so with the same tcp socket so it doesn't seem like a connection issue but I'm not sure what else it could be.  I did mean to add a error type to the api that consumers could use instead of string comparision when deciding if they should retry.

Reply:
Cool yeah not adding to the API at this point makes sense to me, until we know more.

Reply:
Added a sentinel error

Reply:
@cdavis-joy I finally figured out how to get at the actual error code (had to modify libiscsi) and wrote up https://github.com/willgorman/libiscsi-go/issues/2#issuecomment-2512544523 about how we can handle this.  tl;dr is that i think handling the retry caller-side using a sentinel error will be ok-ish for reads (though depending how frequently it happens will add some overhead from re-reads) but we'd want a proper fix before really doing anything with writes